### PR TITLE
Fix webgl regression after WebRender upgrade

### LINKED ISSUE #30367: WebRender upgrade caused a regression in WebGL example TWGL
It seems this patch https://github.com/servo/servo/pull/30323 caused a regression in WebGL example TWGL doesn't run: https://demo.servo.org/experiments/twgl-tunnel/ (again it works revering the change)

The test gets stopped after showing the first image.

WebGL example TWGL stopped

If you resize the window you get some new frames.

WebGL example TWGL with some new image on a smaller window

### PR DESCRIPTION
**Pull request by mukilan**
Sending WebRender an explicit ResourceUpdate message after every buffer swap invalidates the tile caches that depend on the image resource created for the webgl canvas.

The previous WebRender code that Servo was using might have worked with WebGL only accidentally since picture caching was disabled by [default][1] and so the tiles were not cached between frames.

Fixes #30367

[1]: https://github.com/servo/servo/blob/c385b3c9737c17d59cb02e520c3b68b232cb6497/third_party/webrender/webrender/src/renderer.rs#L7073

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #30367 (GitHub issue number if applicable)
- [x] These changes might require tests to catch future regression but this PR doesn't include any. Maybe those can be a separate effort since we want to fix the WebGL demos quickly and look into a proper WebGL test suite later.

### CODE REVIEW

**Review by delan — APPROVED** [review]
Great find! I believe this is exactly what we were missing, because the WebGPU counterpart, Web­GPU­Request­::Swap­Chain­Present, kicks off a Update­Web­Render­Data event that in turn sends an Update­Image transaction. WebGL only did so when the context was resized.

File: components/canvas/webgl_thread.rs [hunk 1]
```diff
@@ -830,6 +831,25 @@ impl WebGLThread {
                 "... rebound framebuffer {}, new back buffer surface is {:?}",
                 framebuffer_object, id
             );
+
+            // Tell Webrender to invalidate the caches for this image
```
Comment (mrobinson) [inline components/canvas/webgl_thread.rs:835]: Sorry for this huge nit:

```suggestion
            // Tell WebRender to invalidate the caches for this image.
```

File: components/canvas/webgl_thread.rs [hunk 2]
```diff
@@ -830,6 +831,25 @@ impl WebGLThread {
...
+                &mut self.webrender_api,
+                self.webrender_doc,
+                size.to_i32(),
+                has_alpha,
+                context_id,
+                info.image_key,
+                image_buffer_kind,
+            );
+
+            debug_assert_eq!(data.gl.get_error(), gl::NO_ERROR);
```
Comment (mrobinson) [inline components/canvas/webgl_thread.rs:852]: Is this new assertion related?

**Review by mrobinson — APPROVED** [review]
Great find @mukilan!

File: components/canvas/webgl_thread.rs (hunk 1 above)
Comment (mukilan) [reply components/canvas/webgl_thread.rs]: It was a lot of work, but I've addressed this huge nit in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f) :rofl:

File: components/canvas/webgl_thread.rs (hunk 2 above)
Comment (mukilan) [reply components/canvas/webgl_thread.rs]: No, you are correct. There are no new OpenGL calls before this, so this assertion is not needed. Removed in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f)

**Comment by mukilan** [comment]
Thanks for the review, @mrobinson and @delan! I've addressed the review comments in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f). I also took the opportunity to refactor the logic shared between resize and swap flows.

**Review by mrobinson — APPROVED** [review]
Thanks so much for this!

File: components/canvas/webgl_thread.rs [hunk 3]
```diff
@@ -908,22 +903,23 @@ impl WebGLThread {
...
+        &mut self,
         context_id: WebGLContextId,
-        image_key: ImageKey,
-        image_buffer_kind: ImageBufferKind,
+        size: Size2D<i32>,
+        has_alpha: bool,
     ) {
-        let descriptor = Self::image_descriptor(size, alpha);
-        let data = Self::external_image_data(context_id, image_buffer_kind);
+        let info = self.cached_context_info.get_mut(&context_id).unwrap();
```
Comment (mrobinson) [inline components/canvas/webgl_thread.rs:914]: Hrm. Does this need to be `get_mut(...)` or can it simply be `get(...)`?

**Review by mrobinson — APPROVED** [review]
Thanks for this fix!

File: components/canvas/webgl_thread.rs (hunk 3 above)
Comment (mukilan) [reply components/canvas/webgl_thread.rs]: Good catch! We can just use 'get'. Fixed in  bce77ef

**Pull request merged**